src/RenameCodeActionProvider.ts

```diff
@@ -0,0 +1,30 @@
+
+import * as vscode from 'vscode';
+import { DefinitionKind } from './INamedDefinition';
+import { PositionContext } from './PositionContext';
+const COMMAND = 'editor.action.rename';
+export class RenameCodeActionProvider implements vscode.CodeActionProvider {
+
+    constructor(private action: (document: vscode.TextDocument, position: vscode.Position) => Promise<PositionContext | undefined>) {
+    }
+
+    public async provideCodeActions(document: vscode.TextDocument, range: vscode.Range): Promise<vscode.CodeAction[] | undefined> {
+        let pc = await this.action(document, range.start);
+        if (!pc) {
+            return;
+        }
+        const referenceSiteInfo = pc.getReferenceSiteInfo(true);
+        if (!referenceSiteInfo || referenceSiteInfo.definition.definitionKind === DefinitionKind.BuiltinFunction) {
```

Comment:
Could you factor out the two usages of definitionKind==DefinitionKind.BuiltinFunction in AzureRMTool.ts plus this one into a new util function util/canRename.ts?  Thanks!  It should return a message ("Built-in functions cannot be renamed." for now), which would be used in the other two places, here you could just check if a message is returned.  Thanks!

Reply:
I did the refactoring as you asked. I think the name of the function is a bit confusing right now: canRename returns undefined if a rename is possible.